storage/pin/pin.go

```diff
@@ -0,0 +1,494 @@
+		pinCounter, err := p.getPinCounterOfChunk(chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
+		if err != nil {
+			log.Error("Error getting pin counter of root hash.", "rootHash", rootHash, "err", err)
+			return nil
+		}
+		fileInfo.pinCounter = pinCounter
+	}
+
+	// Store the pinned files in state DB
+	err = p.savePinnedFile(rootHash, fileInfo)
+	if err != nil {
+		log.Error("Error saving pinned file info to state store.", "rootHash", rootHash, "err", err)
+		return nil
+	}
+
+	log.Debug("File pinned", "Address", rootHash)
+	return nil
+}
+
+// UnPinFiles is used to unpin an already pinned file. It takes the root
+// hash of the file and walks down the merkle tree unpinning all the chunks
+// that are encountered on the way. The pre-requisite is that the file should
+// have been already pinned using the PinFiles function. This function can
+// be called only from an external command.
+func (p *API) UnpinFiles(rootHash string, credentials string) error {
+	addr, err := hex.DecodeString(rootHash)
+	if err != nil {
+		log.Error("Error decoding root hash", "rootHash", rootHash, "err", err)
+		return err
+	}
+
+	fileInfo, err := p.getPinnedFile(rootHash)
+	if err != nil {
+		log.Error("Root hash is not pinned", "rootHash", rootHash, "err", err)
+		return err
+	}
+
+	// Walk the root hash and unpin all the chunks
+	walkerFunction := func(ref storage.Reference) error {
+		chunkAddr := p.removeDecryptionKeyFromChunkHash(ref)
+		err := p.db.Set(context.Background(), chunk.ModeSetUnpin, chunkAddr)
+		if err != nil {
+			log.Error("Could not unpin chunk", "Address", hex.EncodeToString(chunkAddr))
+			return err
+		} else {
+			log.Trace("Unpinning chunk", "Address", hex.EncodeToString(chunkAddr))
+		}
+		return nil
+	}
+	err = p.walkChunksFromRootHash(rootHash, fileInfo.isRaw, credentials, walkerFunction)
+	if err != nil {
+		log.Error("Error walking root hash.", "Hash", rootHash, "err", err)
+		return nil
+	}
+
+	// Delete or Update the state DB
+	pinCounter, err := p.getPinCounterOfChunk(chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
+	if err != nil {
+		err := p.removePinnedFile(rootHash)
+		if err != nil {
+			log.Error("Error unpinning file.", "rootHash", rootHash, "err", err)
+			return nil
+		}
+	} else {
+		fileInfo.pinCounter = pinCounter
+		err = p.savePinnedFile(rootHash, fileInfo)
+		if err != nil {
+			log.Error("Error updating file info to state store.", "rootHash", rootHash, "err", err)
+			return nil
+		}
+	}
+
+	log.Debug("File unpinned", "Address", rootHash)
+	return nil
+}
+
+// ListPinFiles functions logs information of all the files that are pinned
+// in the current local node. It displays the root hash of the pinned file
+// or collection. It also display three vital information's
+//     1) Whether the file is a RAW file or not
+//     2) Size of the pinned file or collection
+//     3) the number of times that particular file or collection is pinned.
+func (p *API) ListPinFiles() (map[string]FileInfo, error) {
```

Comment:
If this function is used only in tests, maybe not to export it?

Reply:
No.. This is exposed outside too. This is the only way through which the node owner will know what files are pinned , their size and their type.

Reply:
OK. In general, exposing such listing functionality should be done with some form of pagination, to prevent very large responses. Maybe that can be considered later if needed.

Reply:
Is there any other command in swarm done with pagination? Just to see and understand. I agree that we should do a pagination... this will be especially useful for dapps which are pinning and pinning service operators like us.

Reply:
I do not think that there is in classical way. But manifests implement a tree that can be considered as a way not to represent all available data at once.

Reply:
ok. I get it.